linera-witty/src/runtime/wasmtime/memory.rs

```diff
@@ -3,44 +3,51 @@
 
 //! How to access the memory of a Wasmtime guest instance.
 
-use super::{super::traits::InstanceWithMemory, EntrypointInstance};
+use super::{super::traits::InstanceWithMemory, EntrypointInstance, ReentrantInstance};
 use crate::{GuestPointer, RuntimeError, RuntimeMemory};
 use std::borrow::Cow;
 use wasmtime::{Extern, Memory};
 
-impl InstanceWithMemory for EntrypointInstance {
-    fn memory_from_export(&self, export: Extern) -> Result<Option<Memory>, RuntimeError> {
-        Ok(match export {
-            Extern::Memory(memory) => Some(memory),
-            _ => None,
-        })
-    }
+macro_rules! impl_memory_traits {
+    ($instance:ty) => {
+        impl InstanceWithMemory for $instance {
+            fn memory_from_export(&self, export: Extern) -> Result<Option<Memory>, RuntimeError> {
+                Ok(match export {
+                    Extern::Memory(memory) => Some(memory),
+                    _ => None,
+                })
+            }
+        }
+
+        impl RuntimeMemory<$instance> for Memory {
+            fn read<'instance>(
+                &self,
+                instance: &'instance $instance,
+                location: GuestPointer,
+                length: u32,
+            ) -> Result<Cow<'instance, [u8]>, RuntimeError> {
+                let start = location.0 as usize;
+                let end = start + length as usize;
+
+                Ok(Cow::Borrowed(&self.data(instance)[start..end]))
+            }
+
+            fn write(
+                &mut self,
+                instance: &mut $instance,
+                location: GuestPointer,
+                bytes: &[u8],
+            ) -> Result<(), RuntimeError> {
+                let start = location.0 as usize;
+                let end = start + bytes.len();
+
+                self.data_mut(instance)[start..end].copy_from_slice(bytes);
+
+                Ok(())
+            }
+        }
+    };
 }
 
-impl RuntimeMemory<EntrypointInstance> for Memory {
-    fn read<'instance>(
-        &self,
-        instance: &'instance EntrypointInstance,
-        location: GuestPointer,
-        length: u32,
-    ) -> Result<Cow<'instance, [u8]>, RuntimeError> {
-        let start = location.0 as usize;
-        let end = start + length as usize;
-
-        Ok(Cow::Borrowed(&self.data(instance)[start..end]))
-    }
-
-    fn write(
-        &mut self,
-        instance: &mut EntrypointInstance,
-        location: GuestPointer,
-        bytes: &[u8],
-    ) -> Result<(), RuntimeError> {
-        let start = location.0 as usize;
-        let end = start + bytes.len();
-
-        self.data_mut(instance)[start..end].copy_from_slice(bytes);
-
-        Ok(())
-    }
-}
+impl_memory_traits!(EntrypointInstance);
```

Comment:
Curious if a trait with a blanket implementation would be able to replace the macros.

Reply:
I'm not sure, because if we did that we could have conflicts between the blanket implementation for Wasmtime and the blanket implementation for Wasmer :thinking: